File: WEEK 2/Day 4/exercise xp gold.py
import json
from urllib.error import HTTPError, URLError
from urllib.parse import urlencode
from urllib.request import urlopen


API_KEY = "hpvZycW22qCjn5cRM1xtWB8NKq4dQ2My"
GIPHY_SEARCH_URL = "https://api.giphy.com/v1/gifs/search"


def fetch_giphy_search(query="hilarious", rating="g", limit=10):
	"""Fetch up to ten search results whose original height is above 100."""
	parameters = urlencode({
		"q": query,
		"rating": rating,
		"api_key": API_KEY,
		"limit": limit,
	})
	url = f"{GIPHY_SEARCH_URL}?{parameters}"

	try:
		with urlopen(url, timeout=15) as response:
			if response.status != 200:
				return {}
			result = json.load(response)
	except (HTTPError, URLError, TimeoutError):
		return {}

	result["data"] = [
		gif for gif in result.get("data", [])
		if int(gif.get("images", {}).get("original", {}).get("height", 0)) > 100
	]
	return result


def main():
	result = fetch_giphy_search()
	print(f"GIFs taller than 100 pixels: {len(result.get('data', []))}")
	for gif in result.get("data", []):
		print(gif.get("url", ""))


if __name__ == "__main__":
	main()

File: WEEK 4/Day 5/Exercise xp gold.js
// =====================================================
// 🌟 EXERCISE 1: GIPHY API #2
// =====================================================

// Giphy API URL
const giphyURL =
  "https://api.giphy.com/v1/gifs/search?q=hilarious&rating=g&api_key=hpvZycW22qCjn5cRM1xtWB8NKq4dQ2My";

// Fetch GIF data
fetch(giphyURL)
  .then(response => {
    // Check response status
    if (!response.ok) {
      throw new Error(`HTTP Error: ${response.status}`);
    }

    return response.json();
  })
  .then(data => {
    console.log("Giphy data:", data);

    // Get all GIFs
    const gifs = data.data;

    // Choose one random GIF
    const randomGif = gifs[Math.floor(Math.random() * gifs.length)];

    // Get the GIF URL from the images object
    const gifURL = randomGif.images.original.url;

    // Create an image element
    const img = document.createElement("img");

    // Add GIF URL
    img.src = gifURL;

    // Add alternative text
    img.alt = "Random Giphy GIF";

    // Add GIF to the webpage
    document.body.appendChild(img);
  })
  .catch(error => {
    console.error("Error fetching GIF:", error);
  });


// =====================================================
// 🌟 EXERCISE 2: ANALYZE #2
// =====================================================

let resolveAfter2Seconds = function () {
  console.log("starting slow promise");

  return new Promise(resolve => {
    setTimeout(function () {
      resolve("slow");
      console.log("slow promise is done");
    }, 2000);
  });
};

let resolveAfter1Second = function () {
  console.log("starting fast promise");

  return new Promise(resolve => {
    setTimeout(function () {
      resolve("fast");
      console.log("fast promise is done");
    }, 1000);
  });
};

let sequentialStart = async function () {
  console.log("==SEQUENTIAL START==");

  const slow = await resolveAfter2Seconds();
  console.log(slow);

  const fast = await resolveAfter1Second();
  console.log(fast);
};

sequentialStart();


// =====================================================
// 🌟 EXERCISE 3: ANALYZE #3
// =====================================================

let concurrentStart = async function () {
  console.log("==CONCURRENT START with await==");

  const slow = resolveAfter2Seconds();
  const fast = resolveAfter1Second();

  console.log(await slow);
  console.log(await fast);
};

// Start after 4 seconds
setTimeout(concurrentStart, 4000);


// =====================================================
// 🌟 EXERCISE 4: MODIFY FETCH WITH ASYNC/AWAIT
// =====================================================

const urls = [
  "https://jsonplaceholder.typicode.com/users",
  "https://jsonplaceholder.typicode.com/posts",
  "https://jsonplaceholder.typicode.com/albums"
];

const getData = async function () {
  try {
    // Fetch all URLs concurrently
    const responses = await Promise.all(
      urls.map(async url => {
        const response = await fetch(url);

        // Check response status
        if (!response.ok) {
          throw new Error(`HTTP Error: ${response.status}`);
        }

        return await response.json();
      })
    );

    // Store the results
    const [users, posts, albums] = responses;

    console.log("users", users);
    console.log("posts", posts);
    console.log("albums", albums);

  } catch (error) {
    console.log("ooooooops");
    console.error(error);
  }
};

getData();
